scheduler/scripts/run_main.py

```diff
@@ -147,39 +152,68 @@ def main(*,
+                    logger.info(f'Retrieving selection for night index {night_idx} '
+                                f'at {site.name} starting at time slot {event_start_time_slot}.')
                     selection = selector.select(night_indices=night_indices,
                                                 sites=frozenset([event.site]),
                                                 starting_time_slots={site: {night_idx: event_start_time_slot
                                                                             for night_idx in night_indices}})
 
-                    # Run the optimizer to get the plans for the first night in the selection.
-                    plans = optimizer.schedule(selection)
-                    night_timeline.add(NightIndex(night_idx),
-                                       site,
-                                       TimeslotIndex(event_start_time_slot),
-                                       event,
-                                       plans[0][site])
-
-                # Perform the time accounting. If the night is done, we execute it to completion.
-                end_timeslot_bounds = None if night_done else {site: TimeslotIndex(event_start_time_slot)}
-                collector.time_accounting(plans[0],
-                                          sites=frozenset({site}),
-                                          end_timeslot_bounds=end_timeslot_bounds)
+                    # Right now the optimizer generates List[Plans], a list of plans indexed by
+                    # every night in the selection. We only want the first one, which corresponds
+                    # to the current night index we are looping over.
+                    logger.info(f'Running optimizer for night index {night_idx} '
+                                f'at {site.name} starting at time slot {event_start_time_slot}.')
+                    plans = optimizer.schedule(selection)[0]
+                    nightly_timeline.add(NightIndex(night_idx),
+                                         site,
+                                         TimeslotIndex(event_start_time_slot),
+                                         event,
+                                         plans[site])
 
         # Piece together the plans for the night to get the overall plans.
+        # This is rather convoluted because of the confusing relationship between Plan, Plans, and NightlyTimeline.
+        # TODO: There appears to be a bug here. See GSCHED-517.
+        logger.info(f'Assembling plans for night index {night_idx}.')
+        night_events = {site: collector.get_night_events(site) for site in collector.sites}
+        final_plans = Plans(night_events, NightIndex(night_idx))
         for site in collector.sites:
-            plans[0][site] = night_timeline.get_final_plan(NightIndex(night_idx), site)
-        overall_plans[night_idx] = plans[0]
+            final_plans[site] = nightly_timeline.get_final_plan(NightIndex(night_idx), site)
```

Comment:
Maybe NightlyTimelime should return the final plans instead of having part of that logic outside.